website/src/models/group.js

```diff
@@ -439,69 +439,55 @@ schema.statics.bossQuest = function bossQuest (user, progress) {
 
 // Remove user from this group
 // TODO this is highly inefficient
-schema.methods.leave = function leaveGroup (user, keep = 'keep-all') {
+schema.methods.leave = async function leaveGroup (user, keep = 'keep-all') {
   let group = this;
 
-  return Q.all([
-    // Remove user from group challenges
-
-    // First find relevant Challenges
-    Challenge.find({
-      _id: {$in: user.challenges}, // Challenges I am in
-      group: group._id, // that belong to the group I am leaving
-    }).then(challenges => {
-      // Update each challenge
-      return Challenge.update(
-        {_id: {$in: _.pluck(challenges, '_id')}},
-        {$pull: {members: user._id}},
-        {multi: true}
-      ).then(() => challenges); // pass `challenges` above to next promise TODO ok to return a non-promise?
-    }).then(challenges => {
-      return Q.all(challenges.map(chal => {
-        let i = user.challenges.indexOf(chal._id);
-        if (i !== -1) user.challenges.splice(i, 1);
-        return user.unlink({cid: chal._id, keep});
-      }));
-    }),
-
-    // Update the group
-    (() => {
-      // If user is the last one in group and group is private, delete it
-      if (group.members.length === 1 && (
-          group.type === 'party' ||
-          group.type === 'guild' && group.privacy === 'private'
-      )) return group.remove();
-
-      let update = {};
-      // otherwise just remove a member TODO create User.methods.removeFromGroup?
-      if (group.type === 'guild') {
-        _.pull(user.guilds, group._id);
-      } else {
-        user.party._id = undefined;
-      }
+  let challenges = await Challenge.find({
+    _id: {$in: user.challenges}, // Challenges I am in
+    group: group._id, // that belong to the group I am leaving
+  });
+
+  await Challenge.update(
+    {_id: {$in: _.pluck(user.challenges, '_id')}},
+    {multi: true}
+  );
 
-      // If the leader is leaving (or if the leader previously left, and this wasn't accounted for)
-      let leader = group.leader;
+  //  remove user from challenges
+  let challengesUserWasRemovedFrom = challenges.map(chal => {
+    let i = user.challenges.indexOf(chal._id);
+    if (i !== -1) user.challenges.splice(i, 1);
+    return user.unlink({cid: chal._id, keep});
```

Comment:
one this branch is up to date with api-v3-groups this should be user.unlinkChallengeTasks(chal._id, keep);